File: src/hlhandler/models/vault.py
"""Vault-related models."""

from dataclasses import dataclass
from decimal import Decimal


@dataclass
class VaultInfo:
    """Information about a vault."""

    address: str
    name: str
    leader: str
    tvl: Decimal
    apr: Decimal
    profit_share: Decimal
    lockup_period: int  # seconds
    is_public: bool
    followers: int = 0
    max_capacity: Decimal | None = None

    @property
    def lockup_hours(self) -> float:
        """Get lockup period in hours."""
        return self.lockup_period / 3600


@dataclass
class VaultPosition:
    """User's position in a vault."""

    vault: str
    vault_name: str
    shares: Decimal
    deposited: Decimal
    current_value: Decimal

    @property
    def pnl(self) -> Decimal:
        """Calculate absolute PnL."""
        return self.current_value - self.deposited

    @property
    def pnl_percent(self) -> Decimal:
        """Calculate PnL percentage."""
        if self.deposited == 0:
            return Decimal("0")
        return (self.pnl / self.deposited) * 100


@dataclass
class VaultDetails:
    """Detailed information about a vault including portfolio."""

    info: VaultInfo
    account_value: Decimal
    positions: list[dict]
    follower_state: dict | None = None  # User's state if following this vault

File: internal/models/vault.go
package models

import "github.com/shopspring/decimal"

// VaultInfo describes a Hyperliquid vault. Mirrors vault.py:VaultInfo.
type VaultInfo struct {
	Address      string
	Name         string
	Leader       string
	TVL          decimal.Decimal
	APR          decimal.Decimal
	ProfitShare  decimal.Decimal
	LockupPeriod int // seconds
	IsPublic     bool
	Followers    int
	MaxCapacity  *decimal.Decimal
}

// LockupHours returns the lockup period in hours.
func (v VaultInfo) LockupHours() float64 { return float64(v.LockupPeriod) / 3600 }

// VaultPosition is a user's stake in a vault. Mirrors vault.py:VaultPosition.
type VaultPosition struct {
	Vault        string
	VaultName    string
	Shares       decimal.Decimal
	Deposited    decimal.Decimal
	CurrentValue decimal.Decimal
}

// Pnl returns the absolute profit/loss.
func (v VaultPosition) Pnl() decimal.Decimal { return v.CurrentValue.Sub(v.Deposited) }

// PnlPercent returns the profit/loss as a percentage of the deposit (0 if the
// deposit is zero).
func (v VaultPosition) PnlPercent() decimal.Decimal {
	if v.Deposited.IsZero() {
		return decimal.Zero
	}
	return v.Pnl().Div(v.Deposited).Mul(decimal.NewFromInt(100))
}

// VaultDetails bundles vault info with the user's portfolio view. Mirrors
// vault.py:VaultDetails.
type VaultDetails struct {
	Info          VaultInfo
	AccountValue  decimal.Decimal
	Positions     []map[string]any
	FollowerState map[string]any // nil if not following
}
